### ISSUE TITLE
different behaviors between using (Small vs Large) Random seeds

### ISSUE THREAD

**Issue by ssydasheng** [google/jax#5050]
Dear jax Developers,

Thanks for your awesome work. I am experimenting with jax for a CIFAR10 classification problem. I found this weird behavior: when I use **small random seeds** (0, 1, ..., 19), the training processes well; when I use **large random seeds** (1M, ..., 9M, 1M+1, ..., 9M+1), the training gets stuck sometimes. I will clarify _"random seeds"_ and _"training gets stuck"_ in the following.

#### Training Gets Stuck
I train each task by 4 epochs, and print the loss on a fixed batch after each epoch. 
A well-working log is like 
```
Init Seed 0 | Epoch 0/3 | Loss 2.3117
Init Seed 0 | Epoch 1/3 | Loss 1.4693
Init Seed 0 | Epoch 2/3 | Loss 1.1867
Init Seed 0 | Epoch 3/3 | Loss 1.0300
```
A get-stuck log is like 
```
Init Seed 1000000 | Epoch 0/3 | Loss 2.3158
Init Seed 1000000 | Epoch 1/3 | Loss 2.3028
Init Seed 1000000 | Epoch 2/3 | Loss 2.3027
Init Seed 1000000 | Epoch 3/3 | Loss 2.3026
```
I am using a CIFAR10 network also used in Zenke et. al., 2017 and Pan et. al., 2020. Firstly I think it weird that the training could get stuck. What's more surprising to me, is the different behaviors between using small and large seeds.

Zenke et. al., 2017, Continual Learning Through Synaptic Intelligence
Pan et. al., 2020, Continual Deep Learning by Functional Regularisation of Memorable Past

#### Random Seed
The **init_seed** is set for _itercount_. The related lines are as follows,

```
@jit
def loss(params, inputs, targets, rng):
     preds = log_softmax(predict(params, inputs, rng=rng))
     targets_one_hot = jnp.asarray(
          targets[:, None] == jnp.arange(preds.shape[-1]), dtype=preds.dtype)
     return -jnp.sum(preds * targets_one_hot, axis=1).mean()

@jit
def update(i, opt_state, inputs, targets):
    params = get_params(opt_state)
    return opt_update(i, grad(loss)(params, inputs, targets, random.PRNGKey(i)), opt_state)

itercount = itertools.count(init_seed)
train_loss = None
for epoch in range(4):
    train_loss = evaluate_task(epoch, get_params(opt_state), train_x[:10000], train_y[:10000])
    print("Init Seed {} | Epoch {}/{} | Loss {:.4f}".format(init_seed, epoch, 3, train_loss))

    for x_batch, y_batch in data_stream_epoch(epoch, train_x, train_y):
         opt_state = update(next(itercount), opt_state, x_batch, y_batch)
```

#### My observations
I run using init_seed: small (0, 1, 2, ...., 19) and large (1M, 2M, ..., 9M, 1M+1, 2M+1, ..., 9M+1) (M is million). For each seed, I run the code for 5 times to counter the GPU non-determinism. For small seeds, all runs worked well. In comparison, for large seeds, training got stuck at once during 5 runs for seeds: 1M, 2M, 8M, 2M+1, 3M+1, 4M+1, 5M+1, 6M+1, 7M+1, 8M+1, . This difference is quite noticeable to me. I understand this type of problem is very difficult to investigate. I just wonder if you might happen to have any thoughts? 

#### Platform and Codes
My jobs are all run using P100. However, since I use slurm for running jobs, I cannot ensure all my runs on the same cpu & gpu. I am using cudnn-10.0-v7.6.3.30, cuda-10.0, jax=0.2.1, jaxlib=0.1.55. 

code: [cifar2.py](https://github.com/google/jax/files/5616973/cifar2.txt)
bash for all runs:  [bash_cifar2.txt](https://github.com/google/jax/files/5616985/bash_cifar2.txt)
logs:  [log_cifar.txt](https://github.com/google/jax/files/5616989/log_cifar.txt)

**Comment by hawkinsp** [comment]
This is actually not related to random seeds at all.

Your code uses a counter based on the random seed (`i`) here:
```
    @jit
    def update(i, opt_state, inputs, targets):
        params = get_params(opt_state)
        return opt_update(i, grad(loss)(params, inputs, targets, random.PRNGKey(i)), opt_state)
```

but you use that count both to:
* seed a PRNG for each minibatch. (I wouldn't do this: you should start with a single PRNGKey and split it each iteration rather than manufacturing a new one from a counter on each iteration. But it's not the bug in this case.)
* as the step number for the ADAM optimizer. This *is* the problem, because the ADAM optimizer uses the step number as part of its bias correction calculation:
https://github.com/google/jax/blob/4378513e15e19776ab5b30881aefa85682b95543/jax/experimental/optimizers.py#L409

and you are effectively starting with a very large step number, which makes the optimizer stop working. This explains why the algorithm works with small "seeds" but fails with large seeds.

Hope that helps!

**Comment by ssydasheng** [comment]
Thanks a lot, this explains it !
